### ISSUE TITLE
DynamicLoader: dlopen does not work

### ISSUE THREAD

**Issue by ADKaster** [SerenityOS/serenity#4642]
An application attempting to use dlopen will get an assert on m_global_symbol_lookup_func.

dlfcn.cpp has its own map of elf objects, separate from the Loader. And, dlopen does not set the required global symbol lookup function like the loader does in map_library.

Ideally, dlopen in dlfcn.cpp would just be a weak reference, and DynamicLoader/main.cpp would create a strong reference to a method that it itself implements, probably one that calls map_library. 

Such an implementation would require #4627 first though :)

**Comment by ADKaster** [comment]
Example from Iovah on IRC

![](https://usercontent.irccloud-cdn.com/file/wvAYd5PO/image.png)

**Comment by alimpfard** [comment]
I was thinking about this, and wondering why we needed loader.so to define `dlopen` and co, is there any reason why `ELF::DynamicLinker` can't provide a `map_library` (and symbol lookup), which would then be called from libc/dlfcn? am I missing some important bit of information here?

**Comment by itamar8910** [comment]
In the current implementation, the dynamic loader comes with its own copy of LibC & LibELF inside of it (which solves the issue of bootstrapping the dependencies of the loader, but makes `Loader.so` quite chubby at ~1MB).

Because of this, the shared libraries data structures from `ELF::DynamicLinker` do not exist in `libc.so` but rather in `Loader.so`.
So the way I see it, the implementation of dlopen etc would need to be inside `Loader.so`.

Regarding the implementation itself, there's a little caveat - I don't think we can currently support defining thread-local-storage data in libraries loaded with `dlopen`, because of the way we implement thread-local-storage.
If the loaded library defines TLS, we would need to call `allocate_tls` again, and i don't really see a way to do that in a thread-safe manner. But anyways I'm not sure supporting this is a must-have feature for us.

**Comment by ADKaster** [comment]
map library needs to be available in the Loader before the app loads LibC. And when the app calls map_library, it needs to reference the same map_library that the loader uses. So the loader needs to take some special care that if there's a relocation for one of the symbols that it owns, the app will reference its implementation. So if we put a fully fledged version in LibC.so, it will be a waste of space, basically. Also, everything itamar said 😁.

**Comment by alimpfard** [comment]
Hmm...in such a case, this seems like a big case of something-I-can't-wrap-my-head-around.
I tried a little and I couldn't even figure out how to make Loader.so substitute something from its own symbol namespace

**Comment by ADKaster** [comment]
A naïve implementation would hook in here 
https://github.com/SerenityOS/serenity/blob/df30b3e54c96f83707bd4fcf6ae158888411efe7/Userland/Libraries/LibELF/DynamicLinker.cpp#L69

And do something like, `if name is one of { "dlopen" "dlsym" "dlerror" "dlclose" } then return getLoaderInternal(name); `

A more performant(?) Solution would create a "fake" dynamic object just for those symbols with a real deal symbol table and gnu hash so that it could go into the g_loaded_objects map and treat those special symbols just like any other STB_WEAK / STB_BIND global symbol.

**Comment by alimpfard** [comment]
yeah, I couldn't write that `getLoaderInternal` thing - is it obvious that this is my first time messing with linking stuff?
I'll maybe try your second suggestion, maybe I could get that to work

**Comment by ADKaster** [comment]
The second option would probably be better because it would allow future work on the userspace emulator to LD_PRELOAD a `libUEStubs.so` that could dlsym(RTLD_NEXT) on all the loader internals (and malloc wink wink) to intercept calls to those guys.


I would expect that the getLoaderInternal function would just turn a pointer to a function pointer wrapped up in a void*/FlatPtr? Same as what you'd get from dlsym. The functions and such in LibC/dlfcn.cpp would definitely need marked weak though. Otherwise the symbol lookup might not have to go to the global lookup function? Haven't looked at that part too closely..

**Comment by alimpfard** [comment]
Pretending for a second that I understand this stuff (lol), why would the first way stop us from doing the `libUEStubs.so` thing? if a strong `dlopen` (for instance) symbol exists, we'll just pick that, otherwise pick the loader's version.
the performance hit can be fixed by just throwing all these things in a hashmap and initialising it only once, further checks will be just a flystring hashmap lookup...

**Comment by ADKaster** [comment]
Well, if the check for special `<dlfcn.h>` symbols happens before we check the g_loaded_objects map for the symbol, then LD_PRELOAD would be pointless. Cuz LD_PRELOAD (when implemented) would just guarantee that that object is loaded first and its STB_GLOBAL symbols are seen before anyone else's.

**Comment by alimpfard** [comment]
so...we can check after...?
if it's still a weak symbol (or doesn't exist) by the time we want to exit from the lookup fn, we can just subst the internal functions?

**Comment by ADKaster** [comment]
Well. Yes of course, could do that 😅. Unlikely anyone will be looking for `dlopen` on their hot paths anyway.

**Comment by alimpfard** [comment]
Further problem, what should happen to a global undefined symbol?
currently our loader just searches the "current" object/library for global symbols, but shouldn't these be resolved against the entire global namespace?

**Comment by ADKaster** [comment]
> currently our loader just searches the "current" object/library for global symbols

That's only true when looking up by name (const char*). DynamicObject:: lookup_symbol(Symbol) looks up in the global registry. So, there should probably be a way to resolve by name globally, yes.

**Comment by ADKaster** [comment]
And to support RTLD_NEXT, we will need a way to explicitly ignore the current object when doing symbol lookup.

**Comment by alimpfard** [comment]
ugh...this is not my cup of tea.
I'll put my patches so far in [here](https://github.com/alimpfard/serenity/tree/dlopen), I don't think I can do this remotely correctly. pretty sure they won't be of any help, but :shrug:

**Comment by alimpfard** [comment]
In a slightly happier mood, I decided to take a look at this again, but now I'm very confused:
```
< CxByte> Hmmm...how does Loader.so resolve symbols defined in a static object?
< CxByte> e.g. libpython3.9.a contains 'exceptions.o', which defines 'PyExc_RuntimeError'. `python3.9` (the
          executable) loads this just fine (?) - how
< CxByte> ?
```
(`import ctypes` in the python port is my test case for now)

I see that `_ctypes.so` has a global relocation for `PyExc_RuntimeError`, but none of the loaded globals have that symbol, so how does it get resolved for `python3.9` (which also needs that symbol)?

**Comment by ADKaster** [comment]
Well, your question doesn't make much sense. The dynamic loader doesn't do anything with static libraries. `foo.a` is not on the radar for the dynamic loader. Any symbols it's working with must be defined either in the main executable, or a library that is DT_NEEDED by the main executable, or a library that's needed by a library that's needed... etc.

**Comment by alimpfard** [comment]
I guess I really don't know what I'm doing...
Here's the situation I find myself in:
- shlibA needs symbol 'X'
- main executable defines symbol 'X'

For whatever reason, this 'X' symbol is not in the loaded globals, and the only difference I can see between this and the ones that _are_ in the globals is the fact that 'X' originally came from a static lib.

**Comment by alimpfard** [comment]
Hmmm...is the loader supposed to ignore the symtab and just look at the dynsym section?

**Comment by ADKaster** [comment]
dlopen now works thanks to the efforts of @gunnarbeutner
